### ISSUE TITLE
Upscaled masks not so accurate (as they can be (!))

### ISSUE THREAD

**Issue by FourierMourier** [ultralytics/ultralytics#4909]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hello there! 💫
lib version: ultralytics 8.0.160
When i was rewriting code from python to C++ I noticed that there's some additional padding from bounding boxes for instance masks
 the third mask after applying the threshold.
<img width="98" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/6e228a2f-4831-4211-bc1a-12f046d2026f">
from the coco128 image using yolov8n-seg.pt:
![000000000312](https://github.com/ultralytics/ultralytics/assets/91980559/33583dc7-5c52-4283-98fa-f0122fb54fbc)

Even when cropping is applied only after upscaling to the input resolution, float values in the rectangles can lead to the omission of one-pixel-width lines at the input image resolution, although this approach may significantly reduce padding,
<img width="130" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/ae5e629a-e531-4cde-92bd-7a8a8eb39761">
potentially requiring additional integer type conversion (see the upper line diff):
<img width="144" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/e7cdaa4e-401f-4da2-bae8-22f6452e555f">

Additionally, upscaling may lead to slower performance in the crop_mask operation, resulting in potential performance differences.
From my point of view, it may be better to apply cropping only after the upscaling.

### Use case

When i was rewriting code from python to C++ I noticed the following: in the ultralytics/models/yolo/segment/predict.py in the `postprocess` method of `SegmentationPredictor` class there's the line:

` masks = ops.process_mask(proto[i], pred[:, 6:], pred[:, :4], img.shape[2:], upsample=True)  # HWC`
where the bounding boxes remove additional info from matrix multiplication between proto and feature vector
function `process_mask` in `ultralytics/utils/ops.py`:
`    masks = crop_mask(masks, downsampled_bboxes)  # CHW`
and the crop_mask itself has the line
`masks * ((r >= x1) * (r < x2) * (c >= y1) * (c < y2))`

The issue arises due to bounding boxes being represented as float values, such as 15.05. This can result in lower bounds for masks (e.g., ((r >= x1))) not including at least the first pixel of the actual mask in the downscaled resolution. After upscaling, there may be padding between the rectangle boundary and the mask's beginning (r >= 15.05 will result in all pixels starting from 16 since r is the torch.Tensor of type int and bbox is the float).
For example, this can be observed with the third mask after applying the threshold.
<img width="98" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/6e228a2f-4831-4211-bc1a-12f046d2026f">
from the coco128 image using yolov8n-seg.pt:
![000000000312](https://github.com/ultralytics/ultralytics/assets/91980559/33583dc7-5c52-4283-98fa-f0122fb54fbc)

To help in the debugging here's some function to get mask as PIL.Image.Image:

```
def get_img_mask(mask: torch.Tensor, bbox: torch.Tensor):
    # downsampled_bboxes[0].round()
    import numpy as np
    from PIL import Image

    rounded = bbox.to(torch.int)
    roi_selected = mask[rounded[1]:rounded[3], rounded[0]:rounded[2]]
    asimage = np.uint8(roi_selected.detach().cpu().numpy()*255.0)
    return Image.fromarray(asimage)
```
   
After line 39 in ultralytics/models/yolo/segment/predict.py in debug mode, you can use the following code to visualize the mask with padding included: get_img_mask(masks[2], pred[2, :4]).show().

My proposed solution is to remove the ROI selection in `crop_mask` and apply it only after upscaling the downscaled mask to the input size. This would look something like this:
`    if upsample:
        masks = F.interpolate(masks[None], shape, mode='bilinear', align_corners=False)[0]  # CHW
    masks = crop_mask(masks, bboxes)  # CHW
    `
    
However, due to float values in the rectangles, one-pixel-width lines may still be missing at the input image resolution (even though much padding can be removed),
<img width="130" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/ae5e629a-e531-4cde-92bd-7a8a8eb39761">
potentially requiring additional integer type conversion (see the upper line diff):
<img width="144" alt="image" src="https://github.com/ultralytics/ultralytics/assets/91980559/e7cdaa4e-401f-4da2-bae8-22f6452e555f">

Additionally, upscaling may lead to slower performance in the crop_mask operation, resulting in potential performance differences.
From my point of view, it may be better to apply cropping only after the upscaling.

### Additional

P.S.: It might be beneficial to first upscale the masks to the original size and then apply the threshold and ROI selection, considering that we plot the masks on the original-sized image rather than input one.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@FourierMourier thanks for the observations! CC'ing @Laughing-q here who did a lot of the Segmentation work. 

If you could submit a PR with your proposed change that would be super helpful and allow us to compare before and after results :)

**Comment by Laughing-q** [comment]
@FourierMourier Hey! Thanks for the observations! However, I think the solution you proposed has already been implemented. You just need to set `retina_masks=True` then it'd use `process_mask_native` function as postprocess which would upscale the masks to the original image size before cropping masks by bounding boxes.
https://github.com/ultralytics/ultralytics/blob/e7401f5872ec0eca6239474a6f3afbd85753fb9d/ultralytics/utils/ops.py#L670-L687

**Comment by FourierMourier** [comment]
> @FourierMourier Hey! Thanks for the observations! However, I think the solution you proposed has already been implemented. You just need to set `retina_masks=True` then it'd use `process_mask_native` function as postprocess which would upscale the masks to the original image size before cropping masks by bounding boxes.
> 
> https://github.com/ultralytics/ultralytics/blob/e7401f5872ec0eca6239474a6f3afbd85753fb9d/ultralytics/utils/ops.py#L670-L687
> 
> 
> EDIT: BTW here's an example of using `retina_masks=True`:
> ```python
> model = YOLO("yolov8n-seg.pt")
> model.predict(retina_masks=True)
> ```

Thanks! It seems i also implemented the same approach when i was rewriting the code :)
But, as it happens, many use default behavior (which is not retina_masks=True) so i'm wondering whether it's the right behavior to crop only after upscaling for `retina_masks=False`  anyway+ I was thinking about int-type conversion and i know this may affect future calculations like `xyxy` argument.
What do you think? Should I try to edit that and submit PR?

**Comment by Laughing-q** [comment]
@FourierMourier Well, retina_masks=False would get faster speed and actually it can satisfy many users so we set retina_masks=False as default behavior. For some cases users got high-resolution images to inference and need more refined masks then we offer retina_masks=True option they can use. I think it's better we just remain current setting.
As for the int-type conversion, do you mean converting all over the boxes to int-type?

**Comment by SilverStarCoder** [comment]
@Laughing-q , related to this, can you confirm if the masks saved to the `predictions.json` file of the `SegmentationValidator`, are of the highest accuracy/quality available? It seems to use `process_mask_upsample` function, which at least according to the docstring of the function, "produces masks of higher quality", but wanted to confirm, because I was not sure exactly what the difference is compared to `process_mask_native`.

**Comment by FourierMourier** [comment]
> @FourierMourier Well, retina_masks=False would get faster speed and actually it can satisfy many users so we set retina_masks=False as default behavior. For some cases users got high-resolution images to inference and need more refined masks then we offer retina_masks=True option they can use. I think it's better we just remain current setting. As for the int-type conversion, do you mean converting all over the boxes to int-type?

Yeah, i'm okay with current retina_masks=True since it does almost the same what i meant and i understand performance issues my possible change would bring. 

For the bounding boxes i mean the target value the user gets as the output from Results is upscaled bboxes to the original size, so maybe masks cropping should occur at this state and in this step bboxes might be converted to int type so there will be no damage to accuracy (but that would also mean changes in several places of the code) - i'm afraid if we convert them too early in the code upscaling will bring new floats...

**Comment by glenn-jocher** [comment]
Hello @FourierMourier! Thanks for your thoughts on this. Your suggestion about cropping the masks after the bounding boxes have been upscaled to the original size makes sense, as this could help maintain accuracy. Regarding converting bounding box coordinates to integers, it can indeed introduce potential inaccuracies due to rounding. Your concerns are valid and we appreciate your thoughts on this. 

As you've mentioned, implementing these changes would involve alterations in several parts of the code, hence bear with us as we carefully consider this option. Your ongoing input is very much appreciated!

Also, as you've suggested, enabling the 'retina_masks' option can help generate more refined masks, particularly for high-resolution images. We affirm this and - as noted - retain the default setting to 'False' for efficiency in many use-cases for our broad user base.
